controllers/apps/transformer_cluster_component.go

```diff
@@ -230,3 +208,353 @@ func copyAndMergeComponent(oldCompObj, newCompObj *appsv1alpha1.Component) *apps
 	}
 	return compObjCopy
 }
+
+const (
+	createOp int = 0
+	deleteOp int = 1
+	updateOp int = 2
+)
+
+func newCompHandler(transCtx *clusterTransformContext, compSpecs map[string]*appsv1alpha1.ClusterComponentSpec,
+	labels, annotations map[string]map[string]string, op int) compConditionalHandler {
+	orders := definedOrders(transCtx, op)
+	if len(orders) == 0 {
+		return newParallelHandler(compSpecs, labels, annotations, op)
+	}
+	return newOrderedHandler(compSpecs, labels, annotations, orders, op)
+}
+
+func definedOrders(transCtx *clusterTransformContext, op int) []string {
+	var (
+		cluster    = transCtx.Cluster
+		clusterDef = transCtx.ClusterDef
+	)
+	if len(cluster.Spec.Topology) != 0 && clusterDef != nil {
+		for _, topology := range clusterDef.Spec.Topologies {
+			if topology.Name == cluster.Spec.Topology {
+				if topology.Orders != nil {
+					switch op {
+					case createOp:
+						return topology.Orders.Provision
+					case deleteOp:
+						return topology.Orders.Terminate
+					case updateOp:
+						return topology.Orders.Update
+					default:
+						panic("runtime error: unknown component op: " + strconv.Itoa(op))
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func newParallelHandler(compSpecs map[string]*appsv1alpha1.ClusterComponentSpec,
+	labels, annotations map[string]map[string]string, op int) compConditionalHandler {
+	switch op {
+	case createOp:
+		return &parallelCreateCompHandler{
+			createCompHandler: createCompHandler{
+				compSpecs:   compSpecs,
+				labels:      labels,
+				annotations: annotations,
+			},
+		}
+	case deleteOp:
+		return &parallelDeleteCompHandler{}
+	case updateOp:
+		return &parallelUpdateCompHandler{
+			updateCompHandler: updateCompHandler{
+				compSpecs:   compSpecs,
+				labels:      labels,
+				annotations: annotations,
+			},
+		}
+	default:
+		panic("runtime error: unknown component op: " + strconv.Itoa(op))
+	}
+}
+
+func newOrderedHandler(compSpecs map[string]*appsv1alpha1.ClusterComponentSpec,
+	labels, annotations map[string]map[string]string, orders []string, op int) compConditionalHandler {
+	switch op {
+	case createOp:
+		return &orderedCreateCompHandler{
+			compOrderedOrder: compOrderedOrder{
+				orders: orders,
+			},
+			compPhasePrecondition: compPhasePrecondition{
+				orders:         orders,
+				expectedPhases: []appsv1alpha1.ClusterComponentPhase{appsv1alpha1.RunningClusterCompPhase},
```

Comment:
why expectedPhases is a list?

Reply:
running is the default strategy currently, and other phases intend to be added later.